client/src/components/PopoverOptionSet/PopoverOptionSet.js

```diff
@@ -0,0 +1,238 @@
+            {i18n._t('PopoverOptionSet.NO_RESULTS', 'No results found')}
+          </div>
+        </div>
+      );
+    }
+
+    return (
+      <div className="popover-option-set__button-container">
+        {buttons.map((button) => (
+          <ButtonComponent
+            className={
+              classNames(
+                button.className,
+                'popover-option-set__button'
+              )
+            }
+            key={button.key}
+            onClick={onButtonClick(button)}
+          >
+            {button.content}
+          </ButtonComponent>
+        ))}
+      </div>
+    );
+  }
+
+  /**
+   * Render the option set popover
+   * @returns {DOMElement}
+   */
+  render() {
+    const { container, extraClass, isOpen, placement, target } = this.props;
+    const popoverClassNames = classNames(
+      'popover-option-set',
+      extraClass
+    );
+
+    return (
+      <Popover
+        className={popoverClassNames}
+        container={container}
+        hideArrow
+        isOpen={isOpen}
+        placement={placement}
+        target={target}
+        toggle={this.handleToggle}
+        onKeyDown={this.handleKeyDown}
+      >
+        {this.renderSearchBox()}
+        {this.renderOptionButtons()}
+      </Popover>
+    );
+  }
+}
+
+const buttonType = PropTypes.shape({
+  key: PropTypes.string.required,
+  content: PropTypes.node.required,
+  className: PropTypes.oneOfType([
+    PropTypes.string,
+    PropTypes.object,
+    PropTypes.arrayOf(PropTypes.string)
+  ]),
+});
+
+PopoverOptionSet.propTypes = {
+  buttons: PropTypes.arrayOf(buttonType).isRequired,
+  // Accepts a function that takes a search term as a first parameter and a set
+  // of buttons to match against that returns a filtered set of buttons
+  // Default search handler assumes button content to be plain text and performs
+  // a simple string.contains check.
+  onSearch: PropTypes.func,
+  // function that accepts a button object & returns an event handler
+  // e.g. (button) => (event) => event.preventDefault() && togglePopover();
+  onButtonClick: PropTypes.func.isRequired,
+  container: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
+  extraClass: PropTypes.oneOfType([PropTypes.string, PropTypes.array, PropTypes.object]),
+  isOpen: PropTypes.bool.isRequired,
+  placement: PropTypes.string,
+  target: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]).isRequired,
+  toggle: PropTypes.func.isRequired,
+  searchPlaceholder: PropTypes.string,
+  disableSearch: PropTypes.bool,
+  ButtonComponent: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
+};
+
+PopoverOptionSet.defaultProps = {
+  searchPlaceholder: i18n._t('PopoverOptionSet.SEARCH_PLACEHOLDER', 'Search'),
+  onSearch: (query, buttons) => buttons.filter(
+    ({ content }) => content.toLowerCase().includes(query.toLowerCase())
+  ),
+  disableSearch: false,
+  ButtonComponent: Button,
+};
+
+export { buttonType };
```

Comment:
We should have a separate place for types - if we don't already it might be a good chance to start

Reply:
Considering that we have lots of components in this module, and types are _usually_ only used by the one component I feel like it makes more sense to keep them here. I can make a type file in the folder for this component if that's nicer?

I really don't think it makes sense to create a new types folder for all components.

Reply:
Ok fair enough. Do you see the number of types for this component expanding? Leaving it where it is will introduce an expectation that it stays where it is (in terms of `import { typeName } from 'path/to/file';`)

Reply:
Ok in the end I just removed the export. Thinking about it, it didn't really make sense as it's not a type that's common across more than just this component. I've in-lined the shape into the proptype definitions now.